Library/Homebrew/diagnostic.rb

```diff
@@ -1111,6 +1111,16 @@ def check_tmpdir
         "TMPDIR #{tmpdir.inspect} doesn't exist." unless tmpdir.nil? || File.directory?(tmpdir)
       end
 
+      def check_tmpdir_executable
+        file = HOMEBREW_TEMP/"homebrew_check_tmpdir_executable"
+        unless system "echo '#!/bin/sh' >#{file} && chmod +x #{file} && #{file}" then <<-EOS.undent
```

Comment:
We should create the file and `chmod` in Ruby (`Pathname#write`, `Pathname#chmod`).